lib/common/socket.c

```diff
@@ -1573,15 +1573,81 @@ void h2o_sliding_counter_stop(h2o_sliding_counter_t *counter, uint64_t now)
 #include "h2o-probes.h"
 #include <sys/stat.h>
 
-static __thread int tracing_map_fd = -1;
-static __thread uint64_t tracing_map_last_attempt = 0;
+static int ebpf_map_create(uint32_t map_type, uint32_t key_size, uint32_t value_size, uint32_t max_entries, const char *map_name)
+{
+    union bpf_attr attr = {
+        .map_type = map_type,
+        .key_size = key_size,
+        .value_size = value_size,
+        .max_entries = max_entries,
+    };
+    strncpy(attr.map_name, map_name, sizeof(attr.map_name));
+    return syscall(__NR_bpf, BPF_MAP_CREATE, &attr, sizeof(attr));
+}
+
+static int ebpf_obj_pin(int bpf_fd, const char *pathname)
+{
+    union bpf_attr attr = {
+        .bpf_fd = (uint32_t)bpf_fd,
+        .pathname = (uint64_t)pathname,
+    };
+    return syscall(__NR_bpf, BPF_OBJ_PIN, &attr, sizeof(attr));
+}
+
+static int ebpf_map_lookup(int fd, const void *key, void *value)
+{
+    union bpf_attr attr = {
+        .map_fd = fd,
+        .key = (uint64_t)key,
+        .value = (uint64_t)value,
+    };
+    return syscall(__NR_bpf, BPF_MAP_LOOKUP_ELEM, &attr, sizeof(attr));
+}
+
+static int ebpf_map_delete(int fd, const void *key)
+{
+    union bpf_attr attr = {
+        .map_fd = fd,
+        .key = (uint64_t)key,
+    };
+    return syscall(__NR_bpf, BPF_MAP_DELETE_ELEM, &attr, sizeof(attr));
+}
+
+int h2o_socket_ebpf_setup(void)
+{
+    if (getuid() != 0) {
+        h2o_error_printf("skipping to set up eBPF maps because bpf(2) requires root privileges\n");
+        return 0;
+    }
 
-static void open_tracing_map(h2o_loop_t *loop)
+    // It creates a pinned BPF object file,
+    // and h2o cannot unlink the file because
+    // h2o drops root privileges after this function
+    // unless its configuration is set to keep the privileges.
+    int fd =
+        ebpf_map_create(BPF_MAP_TYPE_LRU_HASH, sizeof(pid_t), sizeof(uint64_t), H2O_EBPF_RETURN_MAP_SIZE, H2O_EBPF_RETURN_MAP_NAME);
+    if (fd < 0) {
+        h2o_perror("BPF_MAP_CREATE failed");
+        return 0;
+    }
+
+    if (ebpf_obj_pin(fd, H2O_EBPF_RETURN_MAP_PATH) != 0) {
+        h2o_perror("BPF_OBJ_PIN failed");
+        return 0;
+    }
+    close(fd); // each worker thread has its own fd
+    return 1;
+}
```

Comment:
What does the return code of this function indicate? How should the caller handle the result?

I ask the question because I'm confused; if the return code is a boolean indicating success, then I believe that we should be returning 1 when `ebpf_obj_pin` return an error due to a map already been registered (see my comment above regarding the rationale).

This comment also applies to `h2o_socket_ebpf_setup` function defined in the `#else` part.

I do not have a strong preference regarding which, but think that we might want to do either of the following:
* change the return type to `void`, and exit when an error occurs, or
* change the return type to `int`, clarify what the return code means in the doc-comment of the function declaration (i.e. in the header file), and handle that error code at the call-site.

Reply:
That's a good point. 

I was not sure how we can handle the setup errors, but finally, I'd choose the former: the caller needn't handle errors. For our use case, it is an optional setup, and most of the nodes are not required to setup eBPF stuff. Instead, I've tweaked the warning messages for more helpful to users.

Reply:
> For our use case, it is an optional setup, and most of the nodes are not required to setup eBPF stuff.

Negative side of this approach is that we'd miss when things stop working (due to a bug or a deployment issue). Recent example is #2612.

I might argue that h2o should error-exit when it unexpectedly fails to create the return map. If it is hard to tell when it should succeed (or fail), we can have an option to turn on / off the feature. WDYT?

Reply:
Right! It's definitely better to have an option to control this feature. 

How about something like this?

* `ebpf-h2o-return: default | on | off`:
  * Sets up the `h2o_return` eBPF map
  * `default` is "best effort", trying to setup it but ignoring errors. `on` forces to succeed, `off` skips to setup it.

Reply:
Talked with @kazuho elsewhere, and we've agreed with `usdt-selective-tracing: ON | OFF`.